File: internal/cache/cache.go
// Package cache provides an in-memory TTL cache, replicating the
// JS Cache class from the reference project (cache.js).
package cache

import (
	"sync"
	"time"
)

type item struct {
	value     any
	expiresAt time.Time
}

// Cache is a thread-safe in-memory key-value store with TTL expiration.
type Cache struct {
	mu         sync.RWMutex
	store      map[string]item
	defaultTTL time.Duration
}

// New creates a new Cache with the given default TTL.
func New(defaultTTL time.Duration) *Cache {
	if defaultTTL <= 0 {
		defaultTTL = 60 * time.Second
	}
	c := &Cache{
		store:      make(map[string]item),
		defaultTTL: defaultTTL,
	}
	go c.cleanupLoop()
	return c
}

// Get retrieves a value by key. Returns nil, false if expired or missing.
func (c *Cache) Get(key string) (any, bool) {
	c.mu.RLock()
	it, ok := c.store[key]
	c.mu.RUnlock()
	if !ok {
		return nil, false
	}
	if time.Now().After(it.expiresAt) {
		c.mu.Lock()
		delete(c.store, key)
		c.mu.Unlock()
		return nil, false
	}
	return it.value, true
}

// Set stores a value with the default TTL.
func (c *Cache) Set(key string, value any) {
	c.SetTTL(key, value, c.defaultTTL)
}

// SetTTL stores a value with a specific TTL.
func (c *Cache) SetTTL(key string, value any, ttl time.Duration) {
	c.mu.Lock()
	c.store[key] = item{value: value, expiresAt: time.Now().Add(ttl)}
	c.mu.Unlock()
}

// Delete removes a key from the cache.
func (c *Cache) Delete(key string) {
	c.mu.Lock()
	delete(c.store, key)
	c.mu.Unlock()
}

// Clear removes all entries.
func (c *Cache) Clear() {
	c.mu.Lock()
	c.store = make(map[string]item)
	c.mu.Unlock()
}

// GetOrSet returns the cached value, or calls fn to populate it.
func (c *Cache) GetOrSet(key string, fn func() (any, error), ttl time.Duration) (any, error) {
	if v, ok := c.Get(key); ok {
		return v, nil
	}
	v, err := fn()
	if err != nil {
		return nil, err
	}
	c.SetTTL(key, v, ttl)
	return v, nil
}

// cleanupLoop periodically removes expired entries.
func (c *Cache) cleanupLoop() {
	ticker := time.NewTicker(5 * time.Minute)
	defer ticker.Stop()
	for range ticker.C {
		c.mu.Lock()
		now := time.Now()
		for k, v := range c.store {
			if now.After(v.expiresAt) {
				delete(c.store, k)
			}
		}
		c.mu.Unlock()
	}
}

// Stats returns the number of active and expired entries.
func (c *Cache) Stats() (total, active, expired int) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	now := time.Now()
	total = len(c.store)
	for _, v := range c.store {
		if now.After(v.expiresAt) {
			expired++
		} else {
			active++
		}
	}
	return
}

// Cache key constants — mirroring the JS CacheKeys pattern.
const (
	KeyUsers              = "db:users"
	KeyApplications       = "db:applications"
	KeyDemands            = "db:demands"
	KeyEnterprises        = "db:enterprises"
	KeyJobs               = "db:jobs"
	KeyListings           = "db:listings"
	KeyPosts              = "db:posts"
	KeyComments           = "db:comments"
	KeyCertificates       = "db:certificates"
	KeyCourses            = "db:courses"
	KeyInstructors        = "db:instructors"
	KeyPilots             = "db:pilots"
	KeyProducts           = "db:products"
	KeyPolicies           = "db:policies"
	KeyInspections        = "db:inspections"
	KeyLoans              = "db:loans"
	KeyContracts          = "db:contracts"
	KeyLabourOrders       = "db:labour_orders"
	KeyVenues             = "db:venues"
	KeyReviews            = "db:reviews"
	KeyArticles           = "db:articles"
	KeyMessages           = "db:messages"
	KeyEscrowTransactions = "db:escrow_transactions"
	KeyAdminStats         = "stats:admin"
)

// Global cache instance used by the storage layer.
var Global = New(60 * time.Second)

File: microAPP-main/backend/cache.js
/**
 * 缓存模块
 * 提供简单的内存缓存,减少JSON文件读取
 */
class Cache {
  constructor(options = {}) {
    this.store = new Map();
    this.defaultTTL = options.defaultTTL || 60 * 1000; // 默认1分钟
  }

  /**
   * 获取缓存
   */
  get(key) {
    const item = this.store.get(key);

    if (!item) {
      return null;
    }

    // 检查是否过期
    if (Date.now() > item.expiresAt) {
      this.store.delete(key);
      return null;
    }

    return item.value;
  }

  /**
   * 设置缓存
   */
  set(key, value, ttl = this.defaultTTL) {
    this.store.set(key, {
      value,
      expiresAt: Date.now() + ttl
    });
  }

  /**
   * 删除缓存
   */
  delete(key) {
    this.store.delete(key);
  }

  /**
   * 清空所有缓存
   */
  clear() {
    this.store.clear();
  }

  /**
   * 获取或设置(带回调)
   */
  async getOrSet(key, fn, ttl = this.defaultTTL) {
    const cached = this.get(key);

    if (cached !== null) {
      return cached;
    }

    const value = await fn();
    this.set(key, value, ttl);
    return value;
  }

  /**
   * 清理过期缓存
   */
  cleanup() {
    const now = Date.now();
    for (const [key, item] of this.store.entries()) {
      if (now > item.expiresAt) {
        this.store.delete(key);
      }
    }
  }

  /**
   * 获取缓存统计信息
   */
  stats() {
    let expired = 0;
    let active = 0;
    const now = Date.now();

    for (const item of this.store.values()) {
      if (now > item.expiresAt) {
        expired++;
      } else {
        active++;
      }
    }

    return { total: this.store.size, active, expired };
  }
}

// 创建全局缓存实例
const cache = new Cache({ defaultTTL: 60 * 1000 });

// 每5分钟清理一次过期缓存
setInterval(() => cache.cleanup(), 5 * 60 * 1000);

// 缓存键常量
const CacheKeys = {
  USERS: 'db:users',
  APPLICATIONS: 'db:applications',
  CASES: 'db:cases',
  CASE_CATEGORIES: 'db:case_categories',
  SERVICES_CONFIG: 'db:services_config',
  REVIEWS: 'db:reviews',
  MEDICAL_ORDERS: 'db:medical_orders',
  MEDICAL_CERTIFICATIONS: 'db:medical_certifications',
  MEDICAL_PADS: 'db:medical_pads',
  MEDICAL_CONTACTS: 'db:medical_contacts',
  MEDICAL_RATINGS: 'db:medical_ratings',
  MEDICAL_SMS_LOGS: 'db:medical_sms_logs',
  ADMIN_STATS: 'stats:admin',
  USER_INFO: (userId) => `user:${userId}`
};

module.exports = { Cache, cache, CacheKeys };
